database/models/reports.py

```diff
@@ -240,3 +240,31 @@ class CompareComponent(MixinBaseClass, CodecovBaseModel):
     patch_totals = Column(postgresql.JSON)
 
     commit_comparison = relationship(CompareCommit, foreign_keys=[commit_comparison_id])
+
+
+class Test(CodecovBaseModel, MixinBaseClass):
+    __tablename__ = "reports_test"
+    repoid = Column(types.Integer, ForeignKey("repos.repoid"))
+    repository = relationship("Repository", backref=backref("tests"))
+    name = Column(types.String(256), nullable=False)
+    testsuite = Column(types.String(256), nullable=False)
+
+    __table_args__ = (
+        UniqueConstraint(
+            "repoid",
+            "name",
+            "testsuite",
+            name="reports_test_repoid_name_testsuite",
+        ),
+    )
+
+
+class TestInstance(CodecovBaseModel, MixinBaseClass):
+    __tablename__ = "reports_testinstance"
+    test_id = Column(types.Integer, ForeignKey("reports_test.id"))
```

Comment:
hold up, if we change this to a hash of the test suite + test name we might be able to unlock performance improvements

the reason we can't do the db work in the processor task is because we need to synchronize to get/create records of the `Test` model because we need to put its id on the `TestInstance` model. if the ID could be computed from the test suite + test name, then we don't need to synchronize with other tasks to create our `TestInstance` records

as for creating the corresponding `Test` records, we could do `INSERT ... ON CONFLICT DO NOTHING` or `INSERT ... WHERE NOT EXISTS (...)` to attempt inserts in a fire-and-forget way and the database will handle synchronizing internally

Reply:
we would have to hash the env as well but I agree that this may get rid of the issue where we want to use a `returning` clause